File: code/dart/modular_api/test/tracing/w3c_trace_context_propagator_test.dart
import 'package:dartastic_opentelemetry_api/dartastic_opentelemetry_api.dart';
import 'package:modular_api/src/core/tracing/w3c_trace_context_propagator.dart';
import 'package:test/test.dart';

/// Tests for the W3C Trace Context propagator.
///
/// This exists because the Dart OTel **API** does not ship one — it is in the SDK,
/// which core must not depend on (ADR-0005 A2, runbook D26). Python's official
/// package puts an equivalent implementation in its API package, which is both the
/// argument that this belongs at the API layer and the oracle these cases are
/// checked against.
///
/// The format is exact, which is why this stage needs no review gate: 55
/// characters, `00-<32 hex>-<16 hex>-<2 hex>`, version `00` the only one defined,
/// lowercase hex only. Per D4 `tracestate` is carried verbatim and never parsed.
void main() {
  const traceIdHex = '4bf92f3577b34da6a3ce929d0e0e4736';
  const spanIdHex = '00f067aa0ba902b7';
  const sampled = '00-$traceIdHex-$spanIdHex-01';

  setUp(OTelAPI.reset);

  group('W3CTraceContextPropagator', () {
    test('advertises traceparent and tracestate as its fields', () {
      expect(
        const W3CTraceContextPropagator<Map<String, String>>().fields(),
        equals(<String>[traceparentHeader, tracestateHeader]),
      );
    });

    test('extracts trace id, parent span id and the sampled flag', () {
      final spanContext = _extract(sampled).spanContext;

      expect(spanContext, isNotNull);
      expect(spanContext!.traceId.hexString, equals(traceIdHex));
      expect(spanContext.spanId.hexString, equals(spanIdHex));
      expect(spanContext.traceFlags.isSampled, isTrue);
    });

    test('marks an extracted span context as remote', () {
      // Without this a foreign parent is treated as locally created, by both the
      // sampler and the backend.
      expect(_extract(sampled).spanContext!.isRemote, isTrue);
    });

    test('reads the sampled flag from bit 0 of the flags byte', () {
      expect(
        _extract('00-$traceIdHex-$spanIdHex-01').spanContext!.traceFlags.isSampled,
        isTrue,
      );
      expect(
        _extract('00-$traceIdHex-$spanIdHex-00').spanContext!.traceFlags.isSampled,
        isFalse,
      );
      // Unknown flags must not defeat the bit we do understand.
      expect(
        _extract('00-$traceIdHex-$spanIdHex-03').spanContext!.traceFlags.isSampled,
        isTrue,
      );
    });

    group('carries tracestate', () {
      test('verbatim, preserving order', () {
        // D4: we are not a tracing vendor, so we add no entry of our own and we
        // must not normalise, reorder or drop what another vendor set.
        //
        // Asserting the exact round-trip rather than mere presence, because W3C
        // makes tracestate ordering significant — the leftmost entry is the most
        // recently updated system. A `contains` check would pass even if the
        // order were scrambled, which is how this test was first written and what
        // writing the TypeScript mirror exposed.
        const raw = 'rojo=00f067aa0ba902b7,congo=t61rcWkgMzE';
        final spanContext = _extractWith(sampled, tracestate: raw).spanContext;

        expect(spanContext!.traceState?.toString(), equals(raw));
      });

      test('tolerating its absence', () {
        expect(_extract(sampled).spanContext, isNotNull);
      });
    });

    group('leaves the context untouched rather than throwing', () {
      test('when the header is absent', () {
        expect(_extractFrom(const <String, String>{}).spanContext, isNull);
      });

      for (final malformed in <String, String>{
        'unsupported version 01': '01-$traceIdHex-$spanIdHex-01',
        'unsupported version ff': 'ff-$traceIdHex-$spanIdHex-01',
        'short trace id': '00-abc-$spanIdHex-01',
        'short span id': '00-$traceIdHex-abc-01',
        'missing flags field': '00-$traceIdHex-$spanIdHex',
        'extra field': '00-$traceIdHex-$spanIdHex-01-extra',
        'non-hex trace id': '00-${'z' * 32}-$spanIdHex-01',
        'non-hex flags': '00-$traceIdHex-$spanIdHex-zz',
        'uppercase hex': '00-${traceIdHex.toUpperCase()}-$spanIdHex-01',
        'all-zero trace id': '00-${'0' * 32}-$spanIdHex-01',
        'all-zero span id': '00-$traceIdHex-${'0' * 16}-01',
        'empty value': '',
        'unrelated garbage': 'not-a-traceparent',
      }.entries) {
        test('on ${malformed.key}', () {
          expect(_extract(malformed.value).spanContext, isNull);
        });
      }
    });

    test('injects a valid traceparent', () {
      final carrier = _inject(
        OTelAPI.spanContext(
          traceId: OTelAPI.traceIdFrom(traceIdHex),
          spanId: OTelAPI.spanIdFrom(spanIdHex),
          traceFlags: OTelAPI.traceFlags(TraceFlags.SAMPLED_FLAG),
        ),
      );

      expect(carrier[traceparentHeader], equals(sampled));
    });

    test('injects flags 00 for a non-sampled span context', () {
      final carrier = _inject(
        OTelAPI.spanContext(
          traceId: OTelAPI.traceIdFrom(traceIdHex),
          spanId: OTelAPI.spanIdFrom(spanIdHex),
          traceFlags: OTelAPI.traceFlags(TraceFlags.NONE_FLAG),
        ),
      );

      expect(carrier[traceparentHeader], equals('00-$traceIdHex-$spanIdHex-00'));
    });

    test('injects tracestate only when one was received', () {
      final without = _inject(
        OTelAPI.spanContext(
          traceId: OTelAPI.traceIdFrom(traceIdHex),
          spanId: OTelAPI.spanIdFrom(spanIdHex),
        ),
      );
      expect(without.containsKey(tracestateHeader), isFalse);

      final with_ = _inject(
        OTelAPI.spanContext(
          traceId: OTelAPI.traceIdFrom(traceIdHex),
          spanId: OTelAPI.spanIdFrom(spanIdHex),
          traceState: OTelAPI.traceState({'rojo': '00f067aa0ba902b7'}),
        ),
      );
      expect(with_[tracestateHeader], contains('rojo'));
    });

    test('injects nothing when there is no span context to propagate', () {
      final carrier = <String, String>{};
      const W3CTraceContextPropagator<Map<String, String>>()
          .inject(OTelAPI.context(), carrier, _MapSetter(carrier));

      expect(carrier, isEmpty);
    });

    test('round-trips a value it produced', () {
      final original = OTelAPI.spanContext(
        traceId: OTelAPI.traceIdFrom(traceIdHex),
        spanId: OTelAPI.spanIdFrom(spanIdHex),
        traceFlags: OTelAPI.traceFlags(TraceFlags.SAMPLED_FLAG),
      );

      final extracted = _extractFrom(_inject(original)).spanContext;

      expect(extracted!.traceId.hexString, equals(original.traceId.hexString));
      expect(extracted.spanId.hexString, equals(original.spanId.hexString));
      expect(extracted.traceFlags.isSampled, isTrue);
    });
  });
}

Context _extract(String traceparent) =>
    _extractFrom({traceparentHeader: traceparent});

Context _extractWith(String traceparent, {required String tracestate}) =>
    _extractFrom({
      traceparentHeader: traceparent,
      tracestateHeader: tracestate,
    });

Context _extractFrom(Map<String, String> carrier) =>
    const W3CTraceContextPropagator<Map<String, String>>()
        .extract(OTelAPI.context(), carrier, _MapGetter(carrier));

Map<String, String> _inject(SpanContext spanContext) {
  final carrier = <String, String>{};
  const W3CTraceContextPropagator<Map<String, String>>().inject(
    OTelAPI.context().withSpanContext(spanContext),
    carrier,
    _MapSetter(carrier),
  );
  return carrier;
}

/// A carrier reader with HTTP semantics: header names are case-insensitive.
class _MapGetter implements TextMapGetter<String> {
  const _MapGetter(this._carrier);

  final Map<String, String> _carrier;

  @override
  String? get(String key) {
    final wanted = key.toLowerCase();
    for (final entry in _carrier.entries) {
      if (entry.key.toLowerCase() == wanted) return entry.value;
    }
    return null;
  }

  @override
  Iterable<String> keys() => _carrier.keys;
}

class _MapSetter extends TextMapSetter<String> {
  _MapSetter(this._carrier);

  final Map<String, String> _carrier;

  @override
  void set(String key, String value) => _carrier[key] = value;
}

File: code/ts/modular_api/test/tracing/w3cTraceContextPropagator.test.ts
import {
  createTraceState,
  ROOT_CONTEXT,
  TraceFlags,
  trace,
  type Context,
  type SpanContext,
  type TextMapGetter,
  type TextMapSetter,
} from '@opentelemetry/api';
import { describe, expect, it } from 'vitest';

import {
  TRACEPARENT_HEADER,
  TRACESTATE_HEADER,
  W3CTraceContextPropagator,
} from '../../src/core/tracing/w3cTraceContextPropagator';

/**
 * TypeScript mirror of `test/tracing/w3c_trace_context_propagator_test.dart`.
 * Same case names, same expectations; the reference both are checked against is
 * Python's official implementation (runbook D26).
 */
const TRACE_ID = '4bf92f3577b34da6a3ce929d0e0e4736';
const SPAN_ID = '00f067aa0ba902b7';
const SAMPLED = `00-${TRACE_ID}-${SPAN_ID}-01`;

const propagator = new W3CTraceContextPropagator();

/** A carrier reader with HTTP semantics: header names are case-insensitive. */
const headerGetter: TextMapGetter<Record<string, string>> = {
  keys: (carrier) => Object.keys(carrier),
  get: (carrier, key) => {
    const wanted = key.toLowerCase();
    return Object.entries(carrier).find(([name]) => name.toLowerCase() === wanted)?.[1];
  },
};

const headerSetter: TextMapSetter<Record<string, string>> = {
  set: (carrier, key, value) => {
    carrier[key] = value;
  },
};

const extractFrom = (carrier: Record<string, string>): Context =>
  propagator.extract(ROOT_CONTEXT, carrier, headerGetter as TextMapGetter<unknown>);

const extract = (traceparent: string): Context =>
  extractFrom({ [TRACEPARENT_HEADER]: traceparent });

const inject = (spanContext: SpanContext): Record<string, string> => {
  const carrier: Record<string, string> = {};
  propagator.inject(
    trace.setSpanContext(ROOT_CONTEXT, spanContext),
    carrier,
    headerSetter as TextMapSetter<unknown>,
  );
  return carrier;
};

describe('W3CTraceContextPropagator', () => {
  it('advertises traceparent and tracestate as its fields', () => {
    expect(propagator.fields()).toEqual([TRACEPARENT_HEADER, TRACESTATE_HEADER]);
  });

  it('extracts trace id, parent span id and the sampled flag', () => {
    const spanContext = trace.getSpanContext(extract(SAMPLED));

    expect(spanContext).toBeDefined();
    expect(spanContext?.traceId).toBe(TRACE_ID);
    expect(spanContext?.spanId).toBe(SPAN_ID);
    expect(spanContext?.traceFlags).toBe(TraceFlags.SAMPLED);
  });

  it('marks an extracted span context as remote', () => {
    expect(trace.getSpanContext(extract(SAMPLED))?.isRemote).toBe(true);
  });

  it('reads the sampled flag from bit 0 of the flags byte', () => {
    expect(trace.getSpanContext(extract(`00-${TRACE_ID}-${SPAN_ID}-01`))?.traceFlags).toBe(
      TraceFlags.SAMPLED,
    );
    expect(trace.getSpanContext(extract(`00-${TRACE_ID}-${SPAN_ID}-00`))?.traceFlags).toBe(
      TraceFlags.NONE,
    );
    // Unknown flags must not defeat the bit we do understand.
    expect(
      (trace.getSpanContext(extract(`00-${TRACE_ID}-${SPAN_ID}-03`))?.traceFlags ?? 0) &
        TraceFlags.SAMPLED,
    ).toBe(TraceFlags.SAMPLED);
  });

  describe('carries tracestate', () => {
    it('verbatim, preserving order', () => {
      // W3C makes tracestate ordering significant: the leftmost entry is the most
      // recently updated system. This API keeps the raw string, so order survives.
      const raw = 'rojo=00f067aa0ba902b7,congo=t61rcWkgMzE';
      const spanContext = trace.getSpanContext(
        extractFrom({ [TRACEPARENT_HEADER]: SAMPLED, [TRACESTATE_HEADER]: raw }),
      );

      expect(spanContext?.traceState?.serialize()).toBe(raw);
    });

    it('tolerating its absence', () => {
      expect(trace.getSpanContext(extract(SAMPLED))?.traceState).toBeUndefined();
    });
  });

  describe('leaves the context untouched rather than throwing', () => {
    it('when the header is absent', () => {
      expect(trace.getSpanContext(extractFrom({}))).toBeUndefined();
    });

    const malformed: Record<string, string> = {
      'unsupported version 01': `01-${TRACE_ID}-${SPAN_ID}-01`,
      'unsupported version ff': `ff-${TRACE_ID}-${SPAN_ID}-01`,
      'short trace id': `00-abc-${SPAN_ID}-01`,
      'short span id': `00-${TRACE_ID}-abc-01`,
      'missing flags field': `00-${TRACE_ID}-${SPAN_ID}`,
      'extra field': `00-${TRACE_ID}-${SPAN_ID}-01-extra`,
      'non-hex trace id': `00-${'z'.repeat(32)}-${SPAN_ID}-01`,
      'non-hex flags': `00-${TRACE_ID}-${SPAN_ID}-zz`,
      'uppercase hex': `00-${TRACE_ID.toUpperCase()}-${SPAN_ID}-01`,
      'all-zero trace id': `00-${'0'.repeat(32)}-${SPAN_ID}-01`,
      'all-zero span id': `00-${TRACE_ID}-${'0'.repeat(16)}-01`,
      'empty value': '',
      'unrelated garbage': 'not-a-traceparent',
    };

    for (const [label, value] of Object.entries(malformed)) {
      it(`on ${label}`, () => {
        expect(trace.getSpanContext(extract(value))).toBeUndefined();
      });
    }
  });

  it('injects a valid traceparent', () => {
    const carrier = inject({ traceId: TRACE_ID, spanId: SPAN_ID, traceFlags: TraceFlags.SAMPLED });

    expect(carrier[TRACEPARENT_HEADER]).toBe(SAMPLED);
  });

  it('injects flags 00 for a non-sampled span context', () => {
    const carrier = inject({ traceId: TRACE_ID, spanId: SPAN_ID, traceFlags: TraceFlags.NONE });

    expect(carrier[TRACEPARENT_HEADER]).toBe(`00-${TRACE_ID}-${SPAN_ID}-00`);
  });

  it('injects tracestate only when one was received', () => {
    const without = inject({ traceId: TRACE_ID, spanId: SPAN_ID, traceFlags: TraceFlags.NONE });
    expect(without[TRACESTATE_HEADER]).toBeUndefined();

    const withState = inject({
      traceId: TRACE_ID,
      spanId: SPAN_ID,
      traceFlags: TraceFlags.SAMPLED,
      traceState: createTraceState('rojo=00f067aa0ba902b7'),
    });
    expect(withState[TRACESTATE_HEADER]).toBe('rojo=00f067aa0ba902b7');
  });

  it('injects nothing when there is no span context to propagate', () => {
    const carrier: Record<string, string> = {};
    propagator.inject(ROOT_CONTEXT, carrier, headerSetter as TextMapSetter<unknown>);

    expect(carrier).toEqual({});
  });

  it('round-trips a value it produced', () => {
    const original: SpanContext = {
      traceId: TRACE_ID,
      spanId: SPAN_ID,
      traceFlags: TraceFlags.SAMPLED,
    };

    const extracted = trace.getSpanContext(extractFrom(inject(original)));

    expect(extracted?.traceId).toBe(original.traceId);
    expect(extracted?.spanId).toBe(original.spanId);
    expect(extracted?.traceFlags).toBe(TraceFlags.SAMPLED);
  });
});
